### PR TITLE
expose the entire API through the aggregator

### PR DESCRIPTION
**Pull request by deads2k**
This enables the aggregator (unconditionally) and thus exposes CRD.

@danwinship @dcbw per request
@sttts ptal.  I'll link in issues for handling the case where this is active, but we don't have keys.
@bparees does oc cluster up automatically create the keys already?
@sdodson In order to actually *use* the aggregation functionality, you'll need to have the proxy keys available

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by bparees** [comment]
@deads2k if you mean this https://github.com/openshift/origin/blob/master/pkg/bootstrap/docker/openshift/helper.go#L787-L847 then yes?

**Comment by deads2k** [comment]
> @deads2k if you mean this https://github.com/openshift/origin/blob/master/pkg/bootstrap/docker/openshift/helper.go#L787-L847 then yes?

That's the one.  The aggregator bits should become unconditional when using a 3.7 or later image.

**Comment by bparees** [comment]
> That's the one. The aggregator bits should become unconditional when using a 3.7 or later image.

are you telling me oc cluster up is broken right now on 3.7 if you don't specify --service-catalog?

**Comment by deads2k** [comment]
> are you telling me oc cluster up is broken right now on 3.7 if you don't specify --service-catalog?

No.  I'm saying the aggregation feature that we turned off by default in 3.6 becomes unconditionally on 3.7.  If you don't have those cert/key pairs made, you cannot aggregate an external server, but the internal delegation works.

**Comment by bparees** [comment]
> No. I'm saying the aggregation feature that we turned off by default in 3.6 becomes unconditionally on 3.7. If you don't have those cert/key pairs made, you cannot aggregate an external server, but the internal delegation works.

And the SC qualifies as "internal" because it's an internal service?

**Comment by bparees** [comment]
@deads2k so now i think i'm hearing you say i skip all that logic when on 3.7.

**Comment by deads2k** [comment]
> And the SC qualifies as "internal" because it's an internal service?

No. SC is external.

> @deads2k so now i think i'm hearing you say i skip all that logic when on 3.7.

No.  The aggregator key bits are gated on the version or SC and the SC enablement is still fully gated.  Thanks for the link, opened https://github.com/openshift/origin/pull/15355

**Comment by deads2k** [comment]
Ah, it messed with discovery and root URLs, which messed with swagger, which messed the verifies that protect us in the rebase.

@dcbw if you need something to try, this runs as-is, but I'll have to update it tomorrow to pass CI

**Comment by kargakis** [comment]
This is a prerequisite for running prow on top of Openshift.

**Comment by deads2k** [comment]
alright, took a bit of plumbing.  Let's give it anotehr go.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/version.go [hunk 1]
```diff
@@ -19,39 +19,24 @@ package routes
...
 	"k8s.io/apimachinery/pkg/version"
 	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/server/mux"
 )
 
 // Version provides a webservice with version information.
 type Version struct {
-	Version *version.Info
+	Version  *version.Info
+	Delegate http.Handler
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/version.go]: don't see the delegate being used.

File: pkg/cmd/server/api/validation/master.go [hunk 2]
```diff
@@ -229,6 +229,9 @@ func ValidateAggregatorConfig(config api.AggregatorConfig, fldPath *field.Path)
 	validationResults := ValidationResults{}
 
 	validationResults.AddErrors(ValidateCertInfo(config.ProxyClientInfo, false, fldPath.Child("proxyClientInfo"))...)
+	if len(config.ProxyClientInfo.CertFile) == 0 && len(config.ProxyClientInfo.KeyFile) == 0 {
+		validationResults.AddWarnings(field.Invalid(fldPath.Child("proxyClientInfo"), "", "if no client certificate is specified, the aggregator will be unable to proxy to remote server"))
```
Comment (sttts) [inline pkg/cmd/server/api/validation/master.go]: servers

File: pkg/cmd/server/start/start_master.go [hunk 3]
```diff
@@ -403,12 +403,10 @@ func (m *Master) Start() error {
...
-	if len(m.config.AggregatorConfig.ProxyClientInfo.KeyFile) > 0 {
-		// install aggregator types into the scheme so that "normal" RESTOptionsGetters can work for us.
-		// done in Start() prior to doing any other initialization so we don't mutate the scheme after it is being used by clients in other goroutines.
-		// TODO: make scheme threadsafe and do this as part of aggregator config building
-		aggregatorinstall.Install(kapi.GroupFactoryRegistry, kapi.Registry, kapi.Scheme)
-	}
+	// install aggregator types into the scheme so that "normal" RESTOptionsGetters can work for us.
+	// done in Start() prior to doing any other initialization so we don't mutate the scheme after it is being used by clients in other goroutines.
+	// TODO: make scheme threadsafe and do this as part of aggregator config building
+	aggregatorinstall.Install(kapi.GroupFactoryRegistry, kapi.Registry, kapi.Scheme)
```
Comment (sttts) [inline pkg/cmd/server/start/start_master.go]: so I guess the aggregator code itself works nicely without key/cert. haven't checked that.

**Comment by sttts** [comment]
Some nit. Otherwise looks good (if that mode without key+cert works inside the aggregator).

**Comment by deads2k** [comment]
> (if that mode without key+cert works inside the aggregator).

it works enough to unblock the use of CRDs for the networking team.  I don't expect it to work well enough to ship.

**Comment by kargakis** [comment]
> it works enough to unblock the use of CRDs for the networking team. I don't expect it to work well enough to ship.

@smarterclayton can we also use it in api.ci?

**Comment by smarterclayton** [comment]
We need to upgrade api.ci - you and Steve and I need to talk

On Jul 21, 2017, at 8:17 AM, Michail Kargakis <notifications@github.com>
wrote:

it works enough to unblock the use of CRDs for the networking team. I don't
expect it to work well enough to ship.

@smarterclayton <https://github.com/smarterclayton> can we also use it in
api.ci?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/15344#issuecomment-316986065>,
or mute
the thread

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/version.go (hunk 1 above)
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/version.go]: cleaned up

**Comment by deads2k** [comment]
[merge]

**Comment by smarterclayton** [comment]
This looks like the change that made integration take many minutes longer (it's the first jump in integration test times).  I think you're waiting / retrying too slowly / at bad intervals.

**Pull request merged**